### DISCUSSION TITLE
How does Prisma escape strings/identifiers in raw queries?

### DISCUSSION THREAD

**Discussion by molebox** [prisma/prisma#7153] [Q&A] [answered]
I have a script which truncates some tables on a MySQL database. One of these tables uses a reserved word in MySQL - Order. I should be able to escape this by adding quotes around the word, but the transaction still fails as extra quotes get added to the query.

One potential fix is to rename the model (table) and remove the quotes, and this does indeed work, but it doesnt solve the porblem.

The script:

```
  const transactions: PrismaPromise<any>[] = [];
    transactions.push(prisma.$executeRaw`SET FOREIGN_KEY_CHECKS = 0;`)

    for (const { TABLE_NAME } of await prisma.$queryRaw`SELECT TABLE_NAME from information_schema.TABLES WHERE TABLE_SCHEMA = 'tests';`) {
      if (TABLE_NAME !== "_prisma_migrations") {
        try {
          transactions.push(prisma.$executeRaw(
            `TRUNCATE ${TABLE_NAME};`
          ));
        } catch (error) {
          console.log({error})
        }
      }
    }

    transactions.push(prisma.$executeRaw`SET FOREIGN_KEY_CHECKS = 1;`);

    try {
      await prisma.$transaction(transactions);
    } catch (error) {
      console.log({error})
    }
```

The above will fail if our `TABLE_NAME` matches a reserved word.  We can try and escape the reserved word and tell MySQL that its a string by adding some quotes around it:

```
          transactions.push(prisma.$executeRaw(
            `TRUNCATE '${TABLE_NAME}';`
          ));
```

But it seems that under the hood the raw query adds additional quotes and so the string isnt escaped and the transaction fails.

**Comment by janpio** [comment, marked as answer]
Can you isolate the one query that is causing problems? Can you maybe even provide a schema where this applies? 1 model should be enough, right?

**Reply by molebox** [reply]
Sure.

schema, where `Order` is a [reserved word](https://dev.mysql.com/doc/refman/8.0/en/keywords.html)

```
model Order {
  id                  Int            @id @default(autoincrement())
  createdAt     DateTime       @default(now())
  customer      Customer       @relation(fields: [customerId], references: [id])
  customerId   Int
}

model Customer {
  id            Int     @id @default(autoincrement())
  email      String  @unique
  address  String?
  name      String?
  orders    Order[]
}
```

The query fails in the above script at the transaction level, but its in the following where the `TABLE_NAME` is either escaped or not:

```
prisma.$executeRaw(`TRUNCATE ${TABLE_NAME};`)
```

**Reply by pantharshit00** [reply]
`$executeRaw()` doesn't escape anything. It just executes the statement. Only the tagged template literal version 
```js
$executeRaw``
```

escapes the data. It uses https://github.com/blakeembrey/sql-template-tag under the hood and it uses prepared statements: https://dev.mysql.com/doc/refman/8.0/en/sql-prepared-statements.html to do the escaping. 

Now why `'` doesn't work here: Mysql uses `` (backtick) for referencing table names and `'` for strings so that is why it failed. So in order to run this, you need to do the following:

![image](https://user-images.githubusercontent.com/22195362/118882074-641ae980-b911-11eb-9cc4-8fe62229cf86.png)

(sorry for screenshot but markdown wasn't able to properly escape this code block)

**Reply by molebox** [reply]
NICE @pantharshit00 🙌🙌🙌

**Reply by janpio** [reply]
Additional information that Harshit gave me: 
```
$executeRaw`` 
```
does not work with `TRUNCATE` at all (although it is the better one, see docs) because it does not work with prepared statements. I think we talked about this recently in Slack as well. So the `$excecuteRaw()` version needs to be used.
